doc/man1/roarbidir.1
.\" roarbidir.1:

.TH "roarbidir" "1" "January 2010" "RoarAudio" "System User's Manual: roarbidir"

.SH NAME

roarbidir \- Opens a bidirectional stream to RoarAudio's sound server

.SH SYNOPSIS

roarbidir [OPTIONS]... [IN_FILE]

.SH DESCRIPTION

Opens a bidirectional stream to RoarAudio's sound server.
The data is read from stdin or the provideded file and send to stdout.

.SH "OPTIONS"

.TP
\fB--server SERVER\fR
Set server hostname

.TP
\fB--rate   RATE\fR
Set sample rate

.TP
\fB--bits   BITS\fR
Set bits per sample

.TP
\fB--chans  CHANNELS\fR
Set number of channels

.TP
\fB--codec  CODEC\fR
Set the codec

.TP
\fB--help\fR
Show this help

.SH "SEE ALSO"
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3beta3 - ?
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface
	* added server side EsounD emulation
	* added multi homed support to server
	* fixed a long outstanding bug with pa_path_get_filename()

v. 0.2 - Thu May 21 2009 21:00
	* do a some more testings for functions not existing on win32
	* init winsock
	* done a lot fixes in order to get libroar and some clients work on win32
	* done a lot cleanup on configure script

v. 0.2rc2 - Sun May 17 2009 25:44
	* corrected check for GNU peercred
	* added basic logging support to configure
	* fixed plugin configures
	* added and fixed semi-install target on plugins
	* call plugin configures from main configure
	* include plugins in Makefile known to work from configure
	* fixed Makefiles of plugins: added -I../../include, -L../../lib
	* Make use of pkg-config in plugin configures
	* corrected possition of -ls on linker commands in most Makefiles
	* added more explicide linker commands so most things now build
	  with classic BSD make.
	* devided into components: libs, roard, clients, doc
	* add toption to set roard's cycle freq
	* moved a lot things in plugin's configure scripts to a commone one

v. 0.2rc1 - Wed May 13 2009 12:05
	* Started to port to OpenSolaris
	* fixed bug in default based sockets, make HTTP VIO from non default port work
	* got libroarsndio build under OpenBSD 4.5
	* do not depend tests in configure on roaraudio.h: fixes some checks on *BSD
	* renamed member 'fh' of connection struct to '__fh', this breaks
	  all code depending on it so we see how it behaves in case
	  we replace this with a real VIO object in near future
	* started VIO 2popen interface
	* fixed exec() on systems with sizeof(int) != sizeof(char*)
	* do not do unnessesery calls to fdatasync() on sockets
	* link roarvorbis against libogg and libvorbis
	* remove need of GNU/Tar like -z on remote system from send-roaraudio
	* added support for hwmixer flag on OSS and sndio driver
	* fiexed some typos

v. 0.2beta3 - Fri May 01 2009 15:15
	* added esdfilt options to roarfilt, esdplay options to roarcatplay
	* added short options to roarcat/mon/filt as stated in the manpage
	* updated manpages
	* also build libroar{,dsp}.a
	* Changed VIO CTLs in a binary inkompatible way
	* added support for autoconf flag to OSS driver
	* added driver for OpenBSD sndio interface
	* started libroarsndio
	* added code for plugins: audacirus, mplayer, gst10
	* fixed some building problems
	* added roarify and s start for it's manpage
	* added stream flags pause and hwmixer

v. 0.2beta2 - Fri Mar 27 2009 21:20
	* added (dummy) flags autoconf and cleanmeta
	* added streams_ctl() to server
	* added support for primary flag as -oO, support for blocks parameter
	* added (GnuPG) pinentry support.
	* added support for Ogg Vorbis encoder to set meta data :)
	* fixed some (minor) bugs, including segfault in vio
	* added a lot things to the vio interface
	* added gzip and simple OpenPGP support to vio interface
	* added a simple pointer stack.
	* added (OpenSSH) askpass support.
	* added vio pipe support, currectly pipe() and socketpair() based onse.
	* test for OpenSSL
	* added support to VIO open OpenSSL BIOs
	* added support to read a string and convert it into a VIO stack
	* added simple VIO stack as VIO pass object
	* started cleaner VIO Socket implementation

roarclients/roarbidir.c
//roarbidir.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2008
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <roaraudio.h>

#define BUFSIZE 1024

void usage (void) {
 printf("roarbidir [OPTIONS]... [IN_FILE]\n");

 printf("\nOptions:\n\n");

 printf("  --server SERVER    - Set server hostname\n"
        "  --rate   RATE      - Set sample rate\n"
        "  --bits   BITS      - Set bits per sample\n"
        "  --chans  CHANNELS  - Set number of channels\n"
        "  --codec  CODEC     - Set the codec\n"
        "  --help             - Show this help\n"
       );

}

int main (int argc, char * argv[]) {
 int    rate     = 44100;
 int    bits     = 16;
 int    channels = 2;
 int    codec    = ROAR_CODEC_DEFAULT;
 char * server   = NULL;
 char * k;
 int    fh;
 int    i;
 int    in  = -1;
 int    out = -1;
 char buf[BUFSIZE];
 fd_set sl;
 struct timeval tv;
 int max_fh;

 for (i = 1; i < argc; i++) {
  k = argv[i];

  if ( strcmp(k, "--server") == 0 ) {
   server = argv[++i];
  } else if ( strcmp(k, "--rate") == 0 ) {
   rate = atoi(argv[++i]);
  } else if ( strcmp(k, "--bits") == 0 ) {
   bits = atoi(argv[++i]);
  } else if ( strcmp(k, "--channels") == 0 || strcmp(k, "--chans") == 0 ) {
   channels = atoi(argv[++i]);
  } else if ( strcmp(k, "--codec") == 0 ) {
   codec = roar_str2codec(argv[++i]);
  } else if ( strcmp(k, "--help") == 0 ) {
   usage();
   return 0;
  } else if ( in == -1 ) {
   if ( (in = open(k, O_RDONLY, 0644)) == -1 ) {
    fprintf(stderr, "Error: can not open file: %s: %s\n", k, strerror(errno));
    return 1;
   }
  } else {
   fprintf(stderr, "Error: unknown argument: %s\n", k);
   usage();
   return 1;
  }
 }

 if ( (fh = roar_simple_stream(rate, channels, bits, codec, server, ROAR_DIR_BIDIR, "roarbidir")) == -1 ) {
  fprintf(stderr, "Error: can not start playback\n");
  return 1;
 }

 if ( in  == -1 )
  in  = ROAR_STDIN;
 if ( out == -1 )
  out = ROAR_STDOUT;

 max_fh = (in > fh ? in : fh) + 1;
 i      = 1;

 while (i > 0) {
  FD_ZERO(&sl);
  FD_SET(in, &sl);
  FD_SET(fh, &sl);

  tv.tv_sec  = 0;
  tv.tv_usec = 50000;

  if (select(max_fh, &sl, NULL, NULL, &tv) > 0) {
   if ( FD_ISSET(fh, &sl) ) {
    if ( (i = read(fh, buf, BUFSIZE)) == -1 )
     return -1;
    if ( write(out, buf, i) != i )
     return -1;
   }
   if ( FD_ISSET(in, &sl) ) {
    if ( (i = read(in, buf, BUFSIZE)) == -1 )
     return -1;
    if ( write(fh, buf, i) != i )
     return -1;
   }
  }
 }

 roar_simple_close(fh);

 close(in);

 return 0;
}

//ll
